sci-mathematics/sage-7.5.1[doc] - AssertionError: len(context) = 1

After downgrading sphinx to 1.4.4::sage-on-gentoo in response to #450, I encounter a different error while building the documentation:

```
[developer] writing output... [100%] workflows
[developer] Exception occurred:
[developer] File "/usr/lib64/python2.7/site-packages/docutils/writers/_html_base.py", line 671, in depart_document
[developer] assert not self.context, 'len(context) = %s' % len(self.context)
[developer] AssertionError: len(context) = 1
[developer] The full traceback has been saved in /var/tmp/portage/sci-mathematics/sage-7.5.1/temp/sphinx-err-q1SBB8.log, if you want to report 
the issue to the developers.
[developer] Please also report this if it was a user error, so that a better error message can be provided next time.
[developer] A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
Error building the documentation.
Traceback (most recent call last):
  File "sage_setup/docbuild/__main__.py", line 2, in <module>
    main()
  File "/var/tmp/portage/sci-mathematics/sage-7.5.1/work/sage-7.5.1/src-python2_7/sage_setup/docbuild/__init__.py", line 1667, in main
    builder()
  File "/var/tmp/portage/sci-mathematics/sage-7.5.1/work/sage-7.5.1/src-python2_7/sage_setup/docbuild/__init__.py", line 324, in _wrapper
    build_many(build_other_doc, L)
  File "/var/tmp/portage/sci-mathematics/sage-7.5.1/work/sage-7.5.1/src-python2_7/sage_setup/docbuild/__init__.py", line 252, in build_many
    ret = x.get(99999)
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 567, in get
    raise self._value
OSError: [developer] Exception occurred:
```

The mentioned log file contains

```
# Sphinx version: 1.4.4
# Python version: 2.7.12 (CPython)
# Docutils version: 0.13.1 release
# Jinja2 version: 2.9.4
# Last messages:
#   writing output... [ 65%] sagenb/following_latest
#   writing output... [ 69%] sagenb/forking_hell
#   writing output... [ 73%] sagenb/github_development
#   writing output... [ 76%] sagenb/index
#   writing output... [ 80%] sagenb/maintainer_workflow
#   writing output... [ 84%] sagenb/patching
#   writing output... [ 88%] sagenb/set_up_fork
#   writing output... [ 92%] trac
#   writing output... [ 96%] walk_through
#   writing output... [100%] workflows
# Loaded extensions:
#   sphinx.ext.graphviz (1.4.4) from /usr/lib64/python2.7/site-packages/sphinx/ext/graphviz.pyc
#   sphinx.ext.inheritance_diagram (1.4.4) from /usr/lib64/python2.7/site-packages/sphinx/ext/inheritance_diagram.pyc
#   matplotlib.sphinxext.plot_directive (unknown version) from /usr/lib64/python2.7/site-packages/matplotlib/sphinxext/plot_directive.pyc
#   sphinx.ext.extlinks (1.4.4) from /usr/lib64/python2.7/site-packages/sphinx/ext/extlinks.pyc
#   inventory_builder (unknown version) from /var/tmp/portage/sci-mathematics/sage-7.5.1/work/sage-7.5.1/src-python2_7/sage_setup/docbuild/ext/inventory_builder.py
#   sage_autodoc (1.4.4) from /var/tmp/portage/sci-mathematics/sage-7.5.1/work/sage-7.5.1/src-python2_7/sage_setup/docbuild/ext/sage_autodoc.py
#   alabaster (0.7.9) from /usr/lib64/python2.7/site-packages/alabaster/__init__.pyc
#   sphinx.ext.todo (1.4.4) from /usr/lib64/python2.7/site-packages/sphinx/ext/todo.pyc
#   sphinx.ext.mathjax (1.4.4) from /usr/lib64/python2.7/site-packages/sphinx/ext/mathjax.pyc
#   multidocs (unknown version) from /var/tmp/portage/sci-mathematics/sage-7.5.1/work/sage-7.5.1/src-python2_7/sage_setup/docbuild/ext/multidocs.py
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/sphinx/cmdline.py", line 244, in main
    app.build(opts.force_all, filenames)
  File "/usr/lib64/python2.7/site-packages/sphinx/application.py", line 298, in build
    self.builder.build_update()
  File "/usr/lib64/python2.7/site-packages/sphinx/builders/__init__.py", line 251, in build_update
    'out of date' % len(to_build))
  File "/usr/lib64/python2.7/site-packages/sphinx/builders/__init__.py", line 322, in build
    self.write(docnames, list(updated_docnames), method)
  File "/usr/lib64/python2.7/site-packages/sphinx/builders/__init__.py", line 360, in write
    self._write_serial(sorted(docnames), warnings)
  File "/usr/lib64/python2.7/site-packages/sphinx/builders/__init__.py", line 368, in _write_serial
    self.write_doc(docname, doctree)
  File "/usr/lib64/python2.7/site-packages/sphinx/builders/html.py", line 446, in write_doc
    self.docwriter.write(doctree, destination)
  File "/usr/lib64/python2.7/site-packages/docutils/writers/__init__.py", line 80, in write
    self.translate()
  File "/usr/lib64/python2.7/site-packages/sphinx/writers/html.py", line 47, in translate
    self.document.walkabout(visitor)
  File "/usr/lib64/python2.7/site-packages/docutils/nodes.py", line 187, in walkabout
    visitor.dispatch_departure(self)
  File "/usr/lib64/python2.7/site-packages/docutils/nodes.py", line 1895, in dispatch_departure
    return method(node)
  File "/usr/lib64/python2.7/site-packages/docutils/writers/_html_base.py", line 671, in depart_document
    assert not self.context, 'len(context) = %s' % len(self.context)
AssertionError: len(context) = 1
```

That error message appears to indicate some inbalance while traversing the document using a visitor. I couldn't reproduce the problem on the command line because I couldn't get the build started:

```
# cd /var/tmp/portage/sci-mathematics/sage-7.5.1/work/sage-7.5.1/src-python2_7
# python2.7 sage_setup/docbuild/__main__.py --no-pdf-links all html
Traceback (most recent call last):
  File "sage_setup/docbuild/__main__.py", line 1, in <module>
    from sage_setup.docbuild import main
ImportError: No module named sage_setup.docbuild
```

Not sure how the ebuild does this differently; even importing the `environment` didn't make this succeed. So I tried `ebuild …/sage-7.5.1.ebuild compile`. This does seem to do a lot of that doc-building over again, instead of just picking up where the former run left off. By chance I spotted an error message which did not prevent the build from continuing:

```
Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib64/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 326, in _handle_workers
    pool._maintain_pool()
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 230, in _maintain_pool
    self._repopulate_pool()
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 223, in _repopulate_pool
    w.start()
  File "/usr/lib64/python2.7/multiprocessing/process.py", line 130, in start
    self._popen = Popen(self)
  File "/usr/lib64/python2.7/multiprocessing/forking.py", line 121, in __init__
    self.pid = os.fork()
OSError: [Errno 12] Cannot allocate memory
```

The Sage build appears to consume several gigabytes of memory. OK, I did have a Sage computation running which consumed some 9G of memory. I'm not sure whether the fact that this command failed got handled by not creating more threads, or whether this causes part of the documentation to be missing unnoticed. Or whether this is the reason that my sage build appears to be hung now… So I started the build from scratch, and can reproduce the `AssertionError`. Strange!

--- a/dev-python/sphinx/files/sphinx-1.4.1-docutils13.patch
+++ b/dev-python/sphinx/files/sphinx-1.4.1-docutils13.patch
@@ -1,0 +1,34 @@
+# git diff 3adc236114^..7dabcdee91 -- sphinx/writers/html.py
+# and parts of edfdd3ec78 as well in order to make these apply
+
+diff --git a/sphinx/writers/html.py b/sphinx/writers/html.py
+index 7b666aa5..d6bdd833 100644
+--- a/sphinx/writers/html.py
++++ b/sphinx/writers/html.py
+@@ -458,9 +458,9 @@ class HTMLTranslator(BaseTranslator):
+             node['uri'] = posixpath.join(self.builder.imgpath,
+                                          self.builder.images[olduri])
+ 
+-        if node['uri'].lower().endswith('svg') or \
+-           node['uri'].lower().endswith('svgz'):
+-            atts = {'src': node['uri']}
++        uri = node['uri']
++        if uri.lower().endswith(('svg', 'svgz')):
++            atts = {'src': uri}
+             if 'width' in node:
+                 atts['width'] = node['width']
+             if 'height' in node:
+@@ -492,6 +492,13 @@ class HTMLTranslator(BaseTranslator):
+                         node['height'] = str(size[1])
+         BaseTranslator.visit_image(self, node)
+ 
++    # overwritten
++    def depart_image(self, node):
++        if node['uri'].lower().endswith(('svg', 'svgz')):
++            self.body.append(self.context.pop())
++        else:
++            BaseTranslator.depart_image(self, node)
++
+     def visit_toctree(self, node):
+         # this only happens when formatting a toc from env.tocs -- in this
+         # case we don't want to include the subtree

--- a/dev-python/sphinx/sphinx-1.4.4-r1.ebuild
+++ b/dev-python/sphinx/sphinx-1.4.4-r1.ebuild
@@ -45,6 +45,7 @@ DEPEND="${RDEPEND}
 
 PATCHES=(
 	"${FILESDIR}"/${PN}-1.4.1-Makefile.patch
+	"${FILESDIR}"/${PN}-1.4.1-docutils13.patch
 	)
 
 S="${WORKDIR}/${MY_P}"

